File: py_ai/agent/mcts.py
import math
import numpy as np
import torch
import dm_ai_module

class MCTSNode:
    def __init__(self, state, parent=None, action=None):
        self.state = state
        self.parent = parent
        self.action = action # Action that led to this state
        self.children = []
        self.visit_count = 0
        self.value_sum = 0.0
        self.prior = 0.0
        
    def is_expanded(self):
        return len(self.children) > 0

    def value(self):
        if self.visit_count == 0:
            return 0.0
        return self.value_sum / self.visit_count

class MCTS:
    def __init__(self, network, card_db, simulations=100, c_puct=1.0):
        self.network = network
        self.card_db = card_db
        self.simulations = simulations
        self.c_puct = c_puct

    def _fast_forward(self, state):
        while True:
            is_over, _ = dm_ai_module.PhaseManager.check_game_over(state)
            if is_over:
                break
            actions = dm_ai_module.ActionGenerator.generate_legal_actions(state, self.card_db)
            if actions:
                break
            dm_ai_module.PhaseManager.next_phase(state)

    def search(self, root_state):
        root_state_clone = root_state.clone()
        self._fast_forward(root_state_clone)
        root = MCTSNode(root_state_clone)
        
        # Expand root
        self._expand(root)
        
        for _ in range(self.simulations):
            node = root
            
            # Selection
            while node.is_expanded():
                node = self._select_child(node)
                
            # Expansion & Evaluation
            value = self._expand(node)
            
            # Backpropagation
            self._backpropagate(node, value)
            
        return root

    def _select_child(self, node):
        best_score = -float('inf')
        best_child = None
        
        for child in node.children:
            u_score = self.c_puct * child.prior * math.sqrt(node.visit_count) / (1 + child.visit_count)
            q_score = child.value()
            # If active player is different, we might need to flip value?
            # AlphaZero usually assumes value is for the current player.
            # If next state is opponent's turn, the value returned by network for that state is for opponent.
            # So Q should be -Value(child_state).
            # But here we store value_sum.
            # Let's assume standard AlphaZero: Q is mean value.
            # If child state is opponent's turn, the value we get from network is "Probability Opponent Wins".
            # So for us (Parent), it is -Value.
            
            score = q_score + u_score
            if score > best_score:
                best_score = score
                best_child = child
                
        return best_child

    def _expand(self, node):
        # Check game over
        is_over, result = dm_ai_module.PhaseManager.check_game_over(node.state)
        if is_over:
            # Result: 0=None, 1=P1_WIN, 2=P2_WIN, 3=DRAW
            # We need value for the player who just moved (node.parent.state.active_player)
            # Or simply, value for the player at node.state.active_player_id?
            # Usually value is [-1, 1] from perspective of current player.
            # If P1 wins, and current is P1, value = 1.
            # If P1 wins, and current is P2, value = -1.
            
            current_player = node.state.active_player_id
            if result == 3: return 0.0
            if result == 1: return 1.0 if current_player == 0 else -1.0
            if result == 2: return 1.0 if current_player == 1 else -1.0
            return 0.0

        # Generate legal actions
        actions = dm_ai_module.ActionGenerator.generate_legal_actions(node.state, self.card_db)
        
        if not actions:
            # No actions (Pass or Auto-transition?)
            # If empty, maybe we should just step phase?
            # But ActionGenerator should return PASS if allowed.
            # If truly empty, it's a bug or auto-step.
            # Let's assume we step phase and continue expansion?
            # Or treat as terminal?
            # For now, return 0.
            return 0.0

        # Evaluate with Network
        tensor = dm_ai_module.TensorConverter.convert_to_tensor(node.state, node.state.active_player_id)
        tensor_t = torch.tensor(tensor, dtype=torch.float32).unsqueeze(0)
        
        with torch.no_grad():
            policy_logits, value = self.network(tensor_t)
            
        policy = torch.softmax(policy_logits, dim=1).squeeze(0).numpy()
        value = value.item()
        
        # Create children
        for action in actions:
            # Map action to index
            action_idx = dm_ai_module.ActionEncoder.action_to_index(action)
            
            # Clone state
            next_state = node.state.clone()
            # Apply action
            dm_ai_module.EffectResolver.resolve_action(next_state, action, self.card_db)
            # Check if phase changed?
            # If action was PASS, phase changes.
            if action.type == dm_ai_module.ActionType.PASS:
                dm_ai_module.PhaseManager.next_phase(next_state)
            
            # Fast forward through auto-phases
            self._fast_forward(next_state)
            
            child = MCTSNode(next_state, parent=node, action=action)
            
            # Prior probability from policy
            if action_idx >= 0 and action_idx < len(policy):
                child.prior = float(policy[action_idx])
            else:
                child.prior = 0.0 # Should not happen if valid
                
            node.children.append(child)
            
        return value

    def _backpropagate(self, node, value):
        while node is not None:
            node.visit_count += 1
            node.value_sum += value
            value = -value # Flip for opponent
            node = node.parent

File: src/ai/mcts/mcts.cpp
#include "mcts.hpp"
#include "mcts_evaluator.hpp"
#include "engine/actions/intent_generator.hpp"
#include "engine/systems/game_logic_system.hpp"
#include "engine/systems/flow/phase_manager.hpp"
#include "ai/encoders/action_encoder.hpp"
#include "../inference/native_inference.hpp"
#include <cmath>
#include <limits>
#include <random>
#include <algorithm>
#include <iostream>

namespace dm::ai {

    using namespace dm::core;
    using namespace dm::engine;
    using namespace dm::engine::systems;

    MCTSNode::~MCTSNode() {
        // Iterative destruction to avoid stack overflow for deep trees
        std::vector<std::unique_ptr<MCTSNode>> nodes_to_delete;
        // Move children to local vector
        for (auto& child : children) {
            if (child) nodes_to_delete.push_back(std::move(child));
        }
        children.clear();

        while (!nodes_to_delete.empty()) {
            auto node = std::move(nodes_to_delete.back());
            nodes_to_delete.pop_back();

            if (node) {
                for (auto& child : node->children) {
                    if (child) nodes_to_delete.push_back(std::move(child));
                }
                node->children.clear();
            }
        }
    }

    MCTS::MCTS(const std::map<CardID, CardDefinition>& card_db, float c_puct, float dirichlet_alpha, float dirichlet_epsilon, int batch_size, float alpha)
        : card_db_(std::make_shared<std::map<CardID, CardDefinition>>(card_db)), c_puct_(c_puct), dirichlet_alpha_(dirichlet_alpha), dirichlet_epsilon_(dirichlet_epsilon), batch_size_(batch_size), alpha_(alpha) {}

    MCTS::MCTS(std::shared_ptr<const std::map<CardID, CardDefinition>> card_db, float c_puct, float dirichlet_alpha, float dirichlet_epsilon, int batch_size, float alpha)
        : card_db_(card_db), c_puct_(c_puct), dirichlet_alpha_(dirichlet_alpha), dirichlet_epsilon_(dirichlet_epsilon), batch_size_(batch_size), alpha_(alpha) {}

    void MCTS::set_pimc_generator(std::shared_ptr<dm::ai::inference::PimcGenerator> pimc) {
        pimc_generator_ = pimc;
    }

    std::vector<float> MCTS::search(const GameState& root_state, int simulations, BatchEvaluatorCallback evaluator, bool add_noise, float temperature) {
        // PIMC Check: Does the root state have hidden information (ID=0)?
        // We check opponent's hand/shield/deck.
        bool has_hidden_info = false;
        if (pimc_generator_) {
             PlayerID opp_id = 1 - root_state.active_player_id;
             const auto& opp = root_state.players[opp_id];
             // Simple heuristic: if any card in hand is 0, we assume hidden info mode
             if (!opp.hand.empty() && opp.hand[0].card_id == 0) has_hidden_info = true;
             else if (!opp.shield_zone.empty() && opp.shield_zone[0].card_id == 0) has_hidden_info = true;
        }

        if (has_hidden_info) {
             // PIMC Root Parallelization Logic
             std::vector<float> aggregated_policy(ActionEncoder::TOTAL_ACTION_SIZE, 0.0f);
             int samples = std::max(1, pimc_samples_);
             int sims_per_sample = std::max(1, simulations / samples);

             // We need to accumulate policies from different roots
             // Since MCTS::search is stateful (last_root_), we need to be careful.
             // We'll perform multiple searches and aggregate the results.

             std::vector<float> total_policy(ActionEncoder::TOTAL_ACTION_SIZE, 0.0f);

             for (int i = 0; i < samples; ++i) {
                 // 1. Determinize
                 GameState det_state = pimc_generator_->generate_determinized_state(root_state, root_state.active_player_id, i); // Use loop index as seed offset

                 // 2. Run MCTS on this determinized state
                 // Note: This recursively calls search, but now on a state without hidden info.
                 // However, calling search() overwrites last_root_, which is fine as we only care about the policy return.
                 // WARNING: We must ensure the recursive call doesn't trigger PIMC loop again.
                 // Since det_state has filled IDs, has_hidden_info will be false next time.

                 // Disable PIMC generator temporarily for recursive call?
                 // Or rely on the check. The check verifies ID=0. det_state has IDs filled.

                 std::vector<float> pol = this->search(det_state, sims_per_sample, evaluator, add_noise, temperature);

                 for(size_t k=0; k<pol.size(); ++k) {
                     total_policy[k] += pol[k];
                 }
             }

             // Normalize
             for(auto& p : total_policy) {
                 p /= samples;
             }
             return total_policy;
        }

        // Standard MCTS (Perfect Information)

        // 1. Clone and Fast Forward Root
        GameState root_gs = root_state.clone(); // Copy
        PhaseManager::fast_forward(root_gs, *card_db_);
        
        // Explicitly free the old tree before allocating the new one to reduce peak memory usage
        transposition_table_.clear();
        last_root_.reset();
        last_root_ = std::make_unique<MCTSNode>(std::move(root_gs));
        MCTSNode* root = last_root_.get();

        // Initial expansion of root (needs evaluation)
        // Avoid initializer list copying GameState
        std::vector<std::shared_ptr<GameState>> root_batch;
        root_batch.reserve(1);
        root_batch.push_back(std::make_shared<GameState>(root->state.clone()));

        std::pair<std::vector<std::vector<float>>, std::vector<float>> root_eval_res;
        if (dm::ai::inference::NativeInferenceManager::instance().has_model()) {
            dm::ai::mcts::BatchEvaluator be(batch_size_);
            root_eval_res = be.evaluate(root_batch);
        } else {
            if (!evaluator) throw std::runtime_error("No evaluator provided and no native model available");
            root_eval_res = evaluator(root_batch);
        }
        auto &root_policies = root_eval_res.first;
        auto &root_values = root_eval_res.second;
        expand_node(root, root_policies[0]);
        backpropagate(root, root_values[0]);
        
        if (add_noise) {
            add_exploration_noise(root);
        }

        int simulations_finished = 0;
        
        // 3. Simulations Loop
        while (simulations_finished < simulations) {
            std::vector<MCTSNode*> batch_nodes;
            std::vector<std::shared_ptr<GameState>> batch_states;
            
            int current_batch_limit = std::min(batch_size_, simulations - simulations_finished);
            
            for (int i = 0; i < current_batch_limit; ++i) {
                MCTSNode* leaf = select_leaf(root);
                
                bool already_in_batch = false;
                for (auto* n : batch_nodes) {
                    if (n == leaf) {
                        already_in_batch = true;
                        break;
                    }
                }
                
                if (already_in_batch) {
                    revert_virtual_loss(leaf);
                    break;
                }

                GameResult result;
                bool is_over = PhaseManager::check_game_over(leaf->state, result);
                
                if (is_over) {
                    float value = 0.0f;
                    int current_player = leaf->state.active_player_id;
                    if (result == GameResult::DRAW) value = 0.0f;
                    else if (result == GameResult::P1_WIN) value = (current_player == 0) ? 1.0f : -1.0f;
                    else if (result == GameResult::P2_WIN) value = (current_player == 1) ? 1.0f : -1.0f;
                    
                    backpropagate(leaf, value);
                    revert_virtual_loss(leaf);
                    simulations_finished++;
                } else {
                    batch_nodes.push_back(leaf);
                    // Explicit clone wrapped in shared_ptr
                    batch_states.push_back(std::make_shared<GameState>(leaf->state.clone()));
                }
            }
            
            if (batch_nodes.empty()) {
                if (simulations_finished >= simulations) break;
                continue;
            }
            
            std::pair<std::vector<std::vector<float>>, std::vector<float>> eval_res;
            if (dm::ai::inference::NativeInferenceManager::instance().has_model()) {
                dm::ai::mcts::BatchEvaluator be(batch_size_);
                eval_res = be.evaluate(batch_states);
            } else {
                if (!evaluator) throw std::runtime_error("No evaluator provided and no native model available");
                eval_res = evaluator(batch_states);
            }
            auto &policies = eval_res.first;
            auto &values = eval_res.second;
            
            for (size_t i = 0; i < batch_nodes.size(); ++i) {
                MCTSNode* node = batch_nodes[i];
                expand_node(node, policies[i]);

                float val = values[i];
                // Apply dense reward mixing if depth > 0 (node has parent)
                if (node->parent != nullptr) {
                     float dense = node->state.calculate_board_advantage(node->state.active_player_id, *card_db_);
                     val = 0.9f * val + 0.1f * dense;
                }

                backpropagate(node, val);
                revert_virtual_loss(node);
                simulations_finished++;
            }
        }

        // 4. Compute Policy
        std::vector<float> policy(ActionEncoder::TOTAL_ACTION_SIZE, 0.0f);
        
        if (root->children.empty()) return policy;

        if (temperature < 1e-3f) {
            MCTSNode* best_child = nullptr;
            int max_visits = -1;
            for (const auto& child : root->children) {
                if (child->visit_count > max_visits) {
                    max_visits = child->visit_count;
                    best_child = child.get();
                }
            }
            if (best_child) {
                int idx = ActionEncoder::action_to_index(best_child->action_from_parent);
                if (idx >= 0 && idx < (int)policy.size()) {
                    policy[idx] = 1.0f;
                }
            }
        } else {
            float sum_visits_pow = 0.0f;
            std::vector<float> visits_pow;
            visits_pow.reserve(root->children.size());
            
            for (const auto& child : root->children) {
                float v = std::pow(static_cast<float>(child->visit_count), 1.0f / temperature);
                visits_pow.push_back(v);
                sum_visits_pow += v;
            }
            
            for (size_t i = 0; i < root->children.size(); ++i) {
                int idx = ActionEncoder::action_to_index(root->children[i]->action_from_parent);
                if (idx >= 0 && idx < (int)policy.size()) {
                    policy[idx] = visits_pow[i] / sum_visits_pow;
                }
            }
        }

        return policy;
    }

    void MCTS::expand_node(MCTSNode* node, const std::vector<float>& policy_logits) {
        auto actions = IntentGenerator::generate_legal_actions(node->state, *card_db_);
        
        if (actions.empty()) return;

        float sum_exp = 0.0f;
        std::vector<float> priors;
        priors.reserve(actions.size());

        for (const auto& action : actions) {
            int idx = ActionEncoder::action_to_index(action);
            float logit = 0.0f;
            if (idx >= 0 && idx < (int)policy_logits.size()) {
                logit = policy_logits[idx];
            }
            float p = std::exp(logit);
            priors.push_back(p);
            sum_exp += p;
        }

        for (size_t i = 0; i < actions.size(); ++i) {
            const auto& action = actions[i];
            float p = priors[i] / sum_exp;
            
            GameState next_state = node->state.clone();
            // Replaced EffectResolver::resolve_action with GameLogicSystem::resolve_action
            GameLogicSystem::resolve_action(next_state, action, *card_db_);

            if (action.type == PlayerIntent::PASS || action.type == PlayerIntent::MANA_CHARGE) {
                PhaseManager::next_phase(next_state, *card_db_);
            }
            PhaseManager::fast_forward(next_state, *card_db_);
            
            auto child = std::make_unique<MCTSNode>(std::move(next_state)); // next_state is moved

            child->parent = node;
            child->action_from_parent = action;
            child->prior = p;

            size_t hash = child->state.calculate_hash();
            if (MCTSNode* existing = transposition_table_.lookup(hash)) {
                child->visit_count = existing->visit_count;
                child->value_sum = existing->value_sum;
                child->value_squared_sum = existing->value_squared_sum;
            } else {
                transposition_table_.store(hash, child.get());
            }
            
            node->children.push_back(std::move(child));
        }
    }

    MCTSNode* MCTS::select_leaf(MCTSNode* node) {
        while (node->is_expanded()) {
            MCTSNode* best_child = nullptr;
            float best_score = -std::numeric_limits<float>::infinity();

            for (const auto& child : node->children) {
                float q = child->value();
                
                // Risk-Aware Scoring: Score = Q - alpha * sigma
                if (alpha_ > 1e-5f && child->visit_count > 1) {
                    float sigma = child->std_dev();
                    q -= alpha_ * sigma;
                }

                float u = c_puct_ * child->prior * std::sqrt(static_cast<float>(node->visit_count + node->virtual_loss)) / (1.0f + child->visit_count + child->virtual_loss);
                
                float score = q + u;
                if (score > best_score) {
                    best_score = score;
                    best_child = child.get();
                }
            }
            
            if (!best_child) break;
            
            node->virtual_loss++;
            node = best_child;
        }
        
        node->virtual_loss++;
        return node;
    }

    void MCTS::revert_virtual_loss(MCTSNode* node) {
        while (node) {
            node->virtual_loss--;
            node = node->parent;
        }
    }

    void MCTS::backpropagate(MCTSNode* node, float value) {
        while (node) {
            node->visit_count++;

            // Variance accumulation: E[X^2]
            node->value_squared_sum += value * value;

            node->value_sum += value;

            // Invert value only if the active player changes
            if (node->parent && node->parent->state.active_player_id != node->state.active_player_id) {
                value = -value;
            }

            node = node->parent;
        }
    }

    void MCTS::add_exploration_noise(MCTSNode* node) {
        if (node->children.empty()) return;
        
        std::mt19937 rng(std::random_device{}());
        std::gamma_distribution<float> dist(dirichlet_alpha_, 1.0f);
        
        std::vector<float> noise;
        float sum = 0.0f;
        for (size_t i = 0; i < node->children.size(); ++i) {
            float n = dist(rng);
            noise.push_back(n);
            sum += n;
        }
        
        for (size_t i = 0; i < node->children.size(); ++i) {
            node->children[i]->prior = node->children[i]->prior * (1 - dirichlet_epsilon_) + (noise[i] / sum) * dirichlet_epsilon_;
        }
    }

}
